server: adhere to CRI for sandbox stop/remove

runcom opened the pull request:
~~Easily reproducible by stopping a sandbox/container and notice you still have mounts in `mount` output.~~

@mrunalp @sameo @rhatdan PTAL

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

sameo commented on server/container_stop.go:
```diff
@@ -25,6 +25,9 @@ func (s *Server) StopContainer(ctx context.Context, req *pb.StopContainerRequest
 		if err := s.runtime.StopContainer(c, req.Timeout); err != nil {
 			return nil, fmt.Errorf("failed to stop container %s: %v", c.ID(), err)
 		}
+		if err := s.storageRuntimeServer.StopContainer(c.ID()); err != nil {
```
We're now doing this in both `StopContainer` and `RemoveContainer`. The CRI API description seems to indicate that pod sandbox and container resources should be freed at `Stop` time. Do we want to keep removing them from both places ?

runcom replied:
_fixed_

mrunalp replied:
This doesn't seem to be resolved. I still see the call this call in both place :/

runcom replied:
we still need to keep that in both places because remove pod sandbox still needs to stop containers on demand https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L34-L36

Btw, I think sameo was referring to the fact that in container_remove we were still calling StopContainer regardless of status, now in container_remove we call stop_container only if status tells us to do so see here https://github.com/kubernetes-incubator/cri-o/pull/530/files#diff-a7484af4a1e8a5664e3a8829ad1872fcR30

runcom replied:
also the CRI tells us to forcibly remove containers still running here https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L54-L57 so we first stop it and remove it after

sameo commented:
LGTM

mrunalp commented:
will wait for the e2e tests

runcom commented:
corrected the whole PR, the new commit msg is:
```
    we should free resources like IPs/shm/rootfs mounts on sandbox stop,
    not in remove to follow the CRI. The CRI also assures that at least call to
    StopPodSandbox is made before trying to remove the sandbox. That also
    means that this patch updates some tests to correctly stop the sandbox
    before calling remove on it.

    Signed-off-by: Antonio Murdaca <runcom@redhat.com>
```
changed title also.

runcom commented:
@mrunalp @sameo this is good to review now

runcom commented on server/sandbox_stop.go:
```diff
@@ -51,10 +53,38 @@ func (s *Server) StopPodSandbox(ctx context.Context, req *pb.StopPodSandboxReque
+		}
+	}
+
+	if err := s.storageRuntimeServer.StopContainer(sb.id); err != nil {
```
The cleanupDone state variable stinks, I'll remove it and make sure this is idempotent (and add a Stat of the shm mount before trying to umounting it. Hold on the review.

sameo replied:
Fully agree.

runcom commented:
Well at least node-e2e tests ran fine :)

runcom commented:
Updated with more robust process exist check in UpdateStatus

runcom commented:
Working on this again, found some other bugs luckily triggered by kube e2e tests (Note e2e not node-e2e!)

runcom commented:
Good to go again, @mrunalp @sameo PTAL

runcom commented:
retest

runcom commented:
Alright failure in node-e2e isn't a flaky, I'll take a look at it.

runcom commented:
I'm also moving this to v0.5 - the CRI isn't really strong on this topic, `Container|Sandbox[Remove|Stop]` say that we should also stop stuff on remove, that's not really clear, we should open an issue upstream to clarity and tackle this afterwards. Master branch is ok wrt e2e so this isn't really blocking that, moving to v0.5.

TomSweeneyRedHat commented on server/sandbox_remove.go:
```diff
@@ -32,29 +29,16 @@ func (s *Server) RemovePodSandbox(ctx context.Context, req *pb.RemovePodSandboxR
-
+		// At this point we assume containers are stopped _always_
+		// because at least one call to StopPodSandbox is always made
+		// before calling remove
```
Would it be useful to have a validation check here to verify this?   Can/Should we do a quick check on the containers status here and if not stopped return nil and and error?

runcom replied:
that's not true in practice, read https://github.com/kubernetes-incubator/cri-o/pull/530#issuecomment-302958270 - we need to better document this in the CRI proto files. This PR is now on hold.

runcom commented:
@mrunalp @sameo this is finally good to go, as pointed out by @Random-Liu is best to free resources on stop as the call to remove may come too later. PTAL

runcom commented:
@mrunalp @sameo PTAL

sameo commented on server/sandbox_remove.go:
```diff
@@ -74,27 +71,10 @@ func (s *Server) RemovePodSandbox(ctx context.Context, req *pb.RemovePodSandboxR
-		return nil, fmt.Errorf("failed to remove networking namespace for sandbox %s: %v", sb.id, err)
-	}
-
 	s.removeContainer(podInfraContainer)
```
Can we guarantee that kubelet will always do `stop` and then `remove` ? If it goes straight to `remove` we'll be leaking resources here.

runcom replied:
Yeah, that should be guaranteed but we might want to make it clearer in kube API proto file (I mean, nothing to do with this PR right?)

sameo replied:
Agreed, nothing to do with this PR. If you guys are confident k8s/kubelet will never call `remove` before `stop` then this is good to go.
But I believe it would be good for the CRI to explicitly state that and for example expect `RemovePodSandbox` to return an error if the pod is still running.

runcom replied:
Yes we should open an issue in kube to have that stated out clearly

runcom replied:
@Random-Liu wdyt?

feiskyer replied:
@sameo @runcom Yep, this is guaranteed.

TomSweeneyRedHat replied:
Just wondering if a comment hear stating that a stop must be done before the remove is called.  Just as a bread crumb in case someone later is trying to figure things out when memory is leaking and the ordering isn't being held up as it should.  Your call, just a thought.

runcom replied:
That's stated here https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L29-L30 - I won't add it here again as if CRI changes it we need to change it here as well

sameo commented:
LGTM

runcom commented:
@mrunalp PTAL

mrunalp commented:
LGTM
